clippy_lints/src/hash_borrow_str_soundness.rs

```diff
@@ -0,0 +1,82 @@
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    ///
+    ///  This lint is concerned with the soundness of types that implement traits
+    ///  Hash, Borrow<u8> and Borrow<str>:
+    ///  ```
+    /// impl Hash for MyType {
+    /// }
+    ///
+    /// impl Borrow<str> for MyType {
+    /// }
+    ///
+    /// impl Borrow<[u8]> for MyType {
+    /// }
+    /// ```
+    /// and the semantics of Borrow<T> (and BorrowMut). A type is free to borrow as several different
+    /// types.
+    ///
+    /// When providing implementations for Borrow<T>, one should consider whether the different
+    /// implementations should act as facets or representations of the underlying type. Generic code
+    /// typically uses Borrow<T> when it relies on the identical behavior of these additional trait
+    /// implementations. These traits will likely appear as additional trait bounds.
+    ///
+    /// In particular Eq, Ord and Hash must be equivalent for borrowed and owned values:
+    /// `x.borrow() == y.borrow()` should give the same result as `x == y`.
+    ///
+    /// It follows then that the following equivalence must hold:
+    ///  `hash(x) == hash((x as Borrow<[u8]>).borrow()) == hash((x as Borrow<str>).borrow())`
+    ///  Unfortunately it doesn't hold as `hash("abc") != hash("abc".as_bytes())`.
+    #[clippy::version = "1.75.0"]
+    pub HASH_BORROW_STR_SOUNDNESS,
+    correctness,
+    "default lint description"
+}
+
+declare_lint_pass!(HashBorrowStrSoundness => [HASH_BORROW_STR_SOUNDNESS]);
+
+impl LateLintPass<'_> for HashBorrowStrSoundness {
+    /// We are emitting this lint at the Hash impl of a type that implements all
+    /// three of Hash, Borrow<str> and Borrow<[u8]>.
+    ///
+    /// We check that we are in the Hash impl
+    /// Then we check that the type implements Borrow<str> and Borrow<[u8]>
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
+            if let ItemKind::Impl(imp) = item.kind;
+            if let Some(TraitRef {path: Path {span, res, ..}, ..}) = imp.of_trait;
+            if let ty = cx.tcx.type_of(item.owner_id).instantiate_identity();
+            if let Some(hash_id) = cx.tcx.get_diagnostic_item(sym::Hash);
+            if Res::Def(DefKind::Trait, hash_id) == *res;
+            if let Some(borrow_id) = cx.tcx.get_diagnostic_item(sym::Borrow);
+            // since we are in the Hash impl, we don't need to check for that.
+            // we need only to check for Borrow<str> and Borrow<[u8]>
+            if implements_trait(cx, ty, borrow_id, &[cx.tcx.types.str_.into()]);
+            if implements_trait(cx, ty, borrow_id, &[Ty::new_slice(cx.tcx, cx.tcx.types.u8).into()]);
+
+            then {
+                span_lint_and_then(
+                    cx,
+                    HASH_BORROW_STR_SOUNDNESS,
+                    *span,
+                    "implementation of Hash for type that implements Borrow<str> and Borrow<[u8]> is unsound",
```

Comment:
"unsound" has a very specific meaning in Rust and it's concerned with memory safety.
Many Rust devs associate "unsound" with [the ability to cause Undefined Behavior in safe code](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library).
This lint can be triggered in safe code and safe code cannot be unsound.

Unsafe code cannot rely on the correctness of `Hash` or `Borrow` traits, even if it's non-sensical.
I think it would be better to reword this.